post/2017-10-30-how-test-serverless-applications.md

```diff
@@ -0,0 +1,95 @@
+---
+title: How to Test Serverless Applications
+description: "Tips from the Serverless team - how we test serverless applications."
+date: 2017-10-30
+thumbnail: 'https://s3-us-west-2.amazonaws.com/assets.blog.serverless.com/testing-apps.jpg'
+layout: Post
+authors:
+  - EslamHefnawy
+---
+
+Serverless applications are quickly gaining in complexity. Testing is key.
+
+I’ve been building the Serverless Framework for 2 years now, and during that time it's been my focus to create as smooth a testing and debugging experience as possible. In this article, I’ll share some techniques that you can use with the [Serverless Framework](https://www.serverless.com/framework) to test and debug your serverless application during development. 
+
+We'll focus on testing serverless functions, since this is where most of the development happens.
+
+# Unit testing
+
+This one is a gimme. You should always start with unit tests, whether or not your codebase is serverless. Forget about Lambda, the handlers, the events—just organize your codebase in an easy-to-test structure that follows your language’s best practice.
+
+Your handler should always be a thin layer that uses modules out of your code library. If those modules are well-covered with unit tests, then testing the serverless part of your application (i.e., the handlers) will be easy during the integration tests discussed below.
```

Comment:
Could you add a code snippet here to display what you mean? I'm imagining a simple handler that extracts the key elements from an event and then passes it to your real module. Then have a brief piece on testing the module.

Similar in conception to [AWS's example on separating entry point from core logic](http://docs.aws.amazon.com/lambda/latest/dg/best-practices.html#function-code).

Reply:
Agreed! Just added that 👍